File: sdks/typescript/src/evaluators/single-step.ts
import type { ZodType } from 'zod';
import type { ImageAttachment, LLMProvider } from '../providers/index.js';
import { loadImage } from '../features/image-source.js';
import type { EvaluationResult } from '../schemas/index.js';
import type { StageDetail } from '../telemetry/index.js';
import { ConfigurationError, EvaluatorError, wrapProviderError } from '../errors.js';
import { runPreprocessingStep } from '../features/preprocessing.js';
import {
  BaseEvaluator,
  Provider,
  type BaseEvaluatorConfig,
  type EvaluatorMetadata,
} from './base.js';
import { declaredCredentials, type CredentialDeclaringConfig } from './credentials.js';
import { validateInputs, primaryTextField, type DeclaredInputSchema } from './inputs.js';
import { createPromptRenderers } from '../prompts/create-prompts.js';

/**
 * The parts of a contract a single-step evaluator runs on. Everything the flow needs is
 * declared, so the only per-evaluator code left is naming the four artefacts below.
 */
export interface SingleStepContract extends CredentialDeclaringConfig {
  evaluator: {
    id: string;
    stable_id: string;
    id_history: string[];
    name: string;
    description: string;
    /** Required of every contract, so a config.json omitting it fails to compile here. */
    supported_grades: string[];
  };
  steps: Array<{
    id: string;
    model: { provider: string; name: string };
    generation?: { temperature?: number };
    prompt: { placeholders: Record<string, unknown> };
    required_credentials?: string[];
    optional?: boolean;
  }>;
  preprocessing?: Array<{
    id: string;
    implementation: { typescript: { library: string; function: string; post_transform?: { type: string; precision?: number } } };
    required_credentials?: string[];
  }>;
  outcome?: { score: string; reasoning: string };
}

/**
 * An input the model receives as an attached image rather than as text in the prompt.
 *
 * The contract declares the input as a string naming where the image lives (a file path or
 * URL). The shared config schema has no field for this yet, so the declaration sits here
 * until it does; the SDK reads the image in the caller's environment and hands it to the
 * provider as a request attachment, placed ahead of the rendered user prompt.
 */
export interface Attachment {
  /** The input field, as declared in `input_schema.json`. */
  input: string;
  kind: 'image';
}

export interface SingleStepDefinition<TResult> {
  contract: SingleStepContract;
  inputSchema: DeclaredInputSchema;
  outputSchema: ZodType<TResult>;
  /** The contract's `system.txt`, verbatim. Placeholders are substituted per call. */
  systemPrompt: string;
  /** The contract's `user.txt`, verbatim. */
  userPrompt: string;
  /** Inputs attached to the user turn as content parts. Empty for a text-only evaluator. */
  attachments?: readonly Attachment[];
}

/**
 * The class `defineSingleStepEvaluator` returns.
 *
 * Spelled out rather than inferred because an anonymous class cannot carry the base's
 * protected members into the emitted declarations (TS4094).
 */
export interface SingleStepEvaluatorClass<TInput, TResult> {
  new (config: BaseEvaluatorConfig): BaseEvaluator & {
    evaluate(input: TInput): Promise<EvaluationResult<TResult>>;
  };
  readonly metadata: EvaluatorMetadata;
}

/**
 * The step a contract declares under `id`, or a failure naming the one that was expected.
 *
 * Shared because a multi-step evaluator needs the same lookup for each of its steps, and a
 * second copy is a second thing that can disagree about what a missing step means.
 */
export function requireStep<T extends { id: string }>(
  steps: readonly T[],
  id: string,
  evaluatorName: string,
): T {
  const step = steps.find((s) => s.id === id);
  if (!step) {
    throw new Error(`Step "${id}" not found in ${evaluatorName} config.json`);
  }
  return step;
}

/** Step id convention: `evaluate_{slug}`, where slug is the last segment of the id. */
function stepFor(contract: SingleStepContract) {
  const id = `evaluate_${contract.evaluator.id.split('.').pop()}`;
  return requireStep(contract.steps, id, contract.evaluator.name);
}

function vendorOf(step: { model: { provider: string } }, name: string): Provider {
  const vendor = Object.values(Provider).find((p) => p === step.model.provider);
  if (!vendor) {
    throw new Error(
      `Unsupported provider "${step.model.provider}" declared by ${name}. ` +
        `Supported: ${Object.values(Provider).join(', ')}.`,
    );
  }
  return vendor;
}

/**
 * Builds the evaluator class for a contract with one model call.
 *
 * The flow: validate, preprocess, render, call, envelope, telemetry, error wrap. Anything
 * that varies is read from the contract rather than passed in, so an evaluator cannot
 * drift from what its contract declares.
 *
 * @example
 * ```typescript
 * export class PurposeClarityEvaluator extends defineSingleStepEvaluator<
 *   PurposeClarityInput,
 *   PurposeClarityResult
 * >({
 *   contract: CONFIG,
 *   inputSchema: INPUT_SCHEMA,
 *   outputSchema: PurposeClarityOutputSchema,
 *   systemPrompt: SYSTEM_PROMPT,
 *   userPrompt: USER_PROMPT_TEMPLATE,
 * }) {}
 * ```
 */
export function defineSingleStepEvaluator<TInput extends Record<string, string>, TResult>(
  definition: SingleStepDefinition<TResult>,
): SingleStepEvaluatorClass<TInput, TResult> {
  const { contract, inputSchema, outputSchema, systemPrompt, userPrompt } = definition;
  const ATTACHMENTS = definition.attachments ?? [];

  const STEP = stepFor(contract);
  const VENDOR = vendorOf(STEP, contract.evaluator.name);
  const PREPROCESSING = contract.preprocessing ?? [];
  // An attached input is a path or URL, not prose; the primary text is the first field that
  // is neither attached nor an enum.
  const ATTACHED_FIELDS = new Set(ATTACHMENTS.map((a) => a.input));
  const TEXT_FIELD = primaryTextField({
    ...inputSchema,
    properties: Object.fromEntries(
      Object.entries(inputSchema.properties).filter(([name]) => !ATTACHED_FIELDS.has(name)),
    ),
  });
  const PROMPTS = createPromptRenderers(
    systemPrompt,
    userPrompt,
    Object.keys(STEP.prompt.placeholders),
  );

  // The contract names every evaluator "<Thing> Evaluator"; logs read as "<Thing>".
  const LABEL = contract.evaluator.name.replace(/ Evaluator$/, '');

  const METADATA = {
    id: contract.evaluator.id,
    stableId: contract.evaluator.stable_id,
    idHistory: contract.evaluator.id_history,
    name: contract.evaluator.name,
    description: contract.evaluator.description,
    outcome: contract.outcome,
    requiredCredentials: declaredCredentials(contract),
    // What the evaluator targets, which its contract states outright. Deriving it from the
    // grade input instead published `[]` for every evaluator that takes no grade — the
    // feedback family and Grade Level Appropriateness — asserting they target no grades.
    // The accepted set is still the input enum, and `validateInputs` is what enforces it.
    supportedGrades: contract.evaluator.supported_grades as readonly string[],
    defaultProviders: [VENDOR] as const,
  };

  return class SingleStepEvaluator extends BaseEvaluator {
    static readonly metadata = METADATA;

    protected provider: LLMProvider;

    constructor(config: BaseEvaluatorConfig) {
      super(config);
      this.provider = this.createConfiguredProvider(VENDOR, STEP.model.name, this.keyFor(VENDOR, config));
      // Refused here, not at call time: a provider that ignores attachments would otherwise
      // judge the claim with no image and return a confident verdict about nothing.
      if (ATTACHMENTS.length > 0 && !this.provider.supportsAttachments) {
        throw new ConfigurationError(
          `${LABEL} attaches an image to each request, and the configured provider ` +
            `"${this.provider.label}" does not declare support for attachments.`,
        );
      }
    }

    private keyFor(vendor: Provider, config: BaseEvaluatorConfig): string | undefined {
      const keys: Record<Provider, string | undefined> = {
        [Provider.OpenAI]: config.openaiApiKey,
        [Provider.Google]: config.googleApiKey,
        [Provider.Anthropic]: config.anthropicApiKey,
      };
      return keys[vendor];
    }

    async evaluate(input: TInput): Promise<EvaluationResult<TResult>> {
      let text = '';
      let gradeLevel = '';
      const startTime = Date.now();
      const stageDetails: StageDetail[] = [];

      try {
        // Inside the try so a validation failure is telemetered as an error event,
        // and before the inputs are read so a non-object is reported as one.
        validateInputs(input, inputSchema);
        const fields = input as Record<string, string>;
        text = TEXT_FIELD ? fields[TEXT_FIELD] : '';
        gradeLevel = fields.grade_level ?? '';

        this.logger.info(`Starting ${LABEL} evaluation`, {
          evaluator: METADATA.id,
          operation: 'evaluate',
          gradeLevel,
          textLength: text.length,
        });

        // Each declared preprocessing step becomes a prompt input under its own id, so
        // adding one to a contract needs no code here.
        const promptInputs: Record<string, string> = { ...fields };
        for (const step of PREPROCESSING) {
          promptInputs[step.id] = String(runPreprocessingStep(text, step.implementation.typescript));
        }

        // Attached inputs are read here, after validation and before any paid call; the
        // provider places them on the user turn ahead of the text.
        const attachments: ImageAttachment[] = [];
        for (const attachment of ATTACHMENTS) {
          attachments.push(await loadImage(attachment.input, fields[attachment.input]));
        }

        const response = await this.provider.generateStructured({
          messages: [
            { role: 'system', content: PROMPTS.getSystemPrompt(promptInputs) },
            { role: 'user', content: PROMPTS.getUserPrompt(promptInputs) },
          ],
          ...(attachments.length ? { attachments } : {}),
          schema: outputSchema,
          temperature: STEP.generation?.temperature,
        });

        const latencyMs = Date.now() - startTime;
        const tokenUsage = {
          input_tokens: response.usage.inputTokens,
          output_tokens: response.usage.outputTokens,
        };

        stageDetails.push({
          stage: STEP.id,
          provider: this.provider.label,
          latency_ms: response.latencyMs,
          token_usage: tokenUsage,
        });

        const result: EvaluationResult<TResult> = {
          evaluator: METADATA.id,
          result: response.data,
          metadata: {
            model: this.provider.label,
            processingTimeMs: latencyMs,
            tokenUsage: {
              inputTokens: tokenUsage.input_tokens,
              outputTokens: tokenUsage.output_tokens,
            },
          },
        };

        this.sendTelemetry({
          status: 'success',
          latencyMs,
          textLength: text.length,
          gradeLevel,
          provider: this.provider.label,
          tokenUsage,
          metadata: { stage_details: stageDetails },
        }).catch(() => undefined);

        this.logger.info(`${LABEL} evaluation completed successfully`, {
          evaluator: METADATA.id,
          operation: 'evaluate',
          gradeLevel,
          score: METADATA.outcome
            ? (response.data as Record<string, unknown>)[METADATA.outcome.score]
            : undefined,
          processingTimeMs: latencyMs,
        });

        return result;
      } catch (error) {
        const latencyMs = Date.now() - startTime;

        this.logger.error(`${LABEL} evaluation failed`, {
          evaluator: METADATA.id,
          operation: 'evaluate',
          gradeLevel,
          error: error instanceof Error ? error : undefined,
          processingTimeMs: latencyMs,
        });

        const tokenUsage =
          stageDetails.length > 0
            ? {
                input_tokens: stageDetails.reduce((s, d) => s + (d.token_usage?.input_tokens ?? 0), 0),
                output_tokens: stageDetails.reduce((s, d) => s + (d.token_usage?.output_tokens ?? 0), 0),
              }
            : undefined;

        this.sendTelemetry({
          status: 'error',
          latencyMs,
          textLength: text.length,
          gradeLevel,
          provider: this.provider.label,
          tokenUsage,
          errorCode: error instanceof Error ? error.name : 'UnknownError',
          metadata: stageDetails.length > 0 ? { stage_details: stageDetails } : undefined,
        }).catch(() => undefined);

        if (error instanceof EvaluatorError) throw error;
        throw wrapProviderError(error, this.providerContext(this.provider));
      }
    }
  };
}

File: sdks/python/src/learning_commons_evaluators/evaluators/single_step.py
"""The evaluator for a contract with one model call.

The flow — validate, preprocess, render, call, envelope, error wrap — is written once
here. A concrete evaluator is a declaration::

    class PurposeClarityEvaluator(SingleStepEvaluator[PurposeClarityInput, PurposeClarityOutput]):
        contract = load_contract("student_facing_text.ela_reading.purpose_clarity")
        input_model = PurposeClarityInput
        output_model = PurposeClarityOutput

Everything that varies is read from the contract at class creation, so an evaluator
cannot drift from what its contract declares, and a contract this class cannot run fails
at import rather than on the first evaluation. This is the Python form of the TypeScript
SDK's ``defineSingleStepEvaluator`` factory.
"""

from __future__ import annotations

import time
from collections.abc import Mapping
from typing import Any, ClassVar, Generic, TypeVar, cast

from pydantic import BaseModel

from learning_commons_evaluators.contracts.loader import Contract, Preprocessing, Step
from learning_commons_evaluators.errors import ConfigurationError
from learning_commons_evaluators.evaluators.base import BaseEvaluator
from learning_commons_evaluators.evaluators.inputs import primary_text_field, validate_inputs
from learning_commons_evaluators.features.preprocessing import (
    check_implementation,
    format_number,
    run_preprocessing_step,
)
from learning_commons_evaluators.prompts.render import render_prompt
from learning_commons_evaluators.providers import (
    LLMProvider,
    Message,
    Provider,
    call_with_resampling,
    provider_context,
)
from learning_commons_evaluators.schemas.evaluator import (
    EvaluationMetadata,
    EvaluationResult,
    EvaluationTokenUsage,
)
from learning_commons_evaluators.schemas.metadata import EvaluatorMetadata

InputT = TypeVar("InputT", bound=BaseModel)
OutputT = TypeVar("OutputT", bound=BaseModel)


def step_for(contract: Contract) -> Step:
    """The step the convention names: ``evaluate_{slug}``, where slug is the id's last segment."""
    return contract.step(f"evaluate_{contract.evaluator.slug}")


class SingleStepEvaluator(BaseEvaluator, Generic[InputT, OutputT]):
    """Base for every evaluator whose contract declares one LLM step."""

    #: The declarations a concrete class makes, alongside ``contract`` from the base.
    input_model: ClassVar[type[BaseModel]]
    output_model: ClassVar[type[BaseModel]]

    # Resolved from the contract at class creation.
    _step: ClassVar[Step]
    _vendor: ClassVar[Provider]
    _preprocessing: ClassVar[tuple[Preprocessing, ...]]
    _text_field: ClassVar[str | None]

    def __init_subclass__(cls, **kwargs: Any) -> None:
        super().__init_subclass__(**kwargs)
        if not hasattr(cls, "contract"):
            return  # an abstract intermediate; the declaring subclass resolves
        contract = cls.contract
        name = contract.evaluator.name

        # Exactly one step, and not an optional one: this base runs the step the convention
        # names and nothing else, so any other step would be silently skipped while the
        # metadata and credential checks described the whole contract.
        if len(contract.steps) != 1 or contract.steps[0].optional:
            raise ValueError(
                f"{name} config.json declares {len(contract.steps)} step(s); SingleStepEvaluator "
                "runs exactly one non-optional step. Use the multi-step base."
            )
        step = step_for(contract)
        if step.type != "llm" or step.prompt is None or step.model is None:
            raise ValueError(f'Step "{step.id}" in {name} config.json is not an LLM step.')
        for placeholder, declared in step.prompt.placeholders.items():
            if declared.source.startswith("steps."):
                raise ValueError(
                    f'Placeholder "{placeholder}" in {name} reads another step; '
                    "a single-step evaluator has none. Use the multi-step base."
                )

        for entry in contract.preprocessing:
            if entry.type != "computation" or entry.python is None:
                raise ValueError(
                    f'Preprocessing "{entry.id}" in {name} config.json declares no Python '
                    "computation this evaluator can run."
                )
            # A library or transform this SDK lacks is our gap, not the provider's: fail
            # here, at import, rather than inside the evaluation's error boundary.
            try:
                check_implementation(entry.python)
            except NotImplementedError as gap:
                raise NotImplementedError(
                    f'Preprocessing "{entry.id}" in {name} config.json: {gap}'
                ) from None

        cls._step = step
        cls._vendor = step.model.provider
        cls._preprocessing = tuple(contract.preprocessing)
        cls._text_field = primary_text_field(contract.input_schema)
        cls.metadata = EvaluatorMetadata.from_contract(contract, (step.model.provider,))

    def __init__(self, config: Any = None, /, **fields: Any) -> None:
        super().__init__(config, **fields)
        assert self._step.model is not None  # established at class creation
        self.provider: LLMProvider = self._create_configured_provider(
            self._vendor, self._step.model.name
        )

    async def evaluate(
        self, input: InputT | None = None, /, **fields: Any
    ) -> EvaluationResult[OutputT]:
        """Evaluate the contract's inputs, passed as the typed input model or by name.

        :raises InputValidationError: an input is missing, unknown, or outside its schema.
        :raises ConfigurationError: the provider rejected the configured model id, or a
            required placeholder has no value from the source the contract names.
        :raises DependencyError: the provider call failed (``AuthenticationError``,
            ``RateLimitError``, ``NetworkError``, ``RequestTimeoutError``, ``LLMProviderError``).
        :raises LLMOutputProcessingError: the model's response failed its output schema
            after ``max_retries`` immediate resamples.
        """
        start = time.perf_counter()
        context = {"evaluator": self.metadata.id, "operation": "evaluate"}
        grade_level = ""
        # Argument shape is a programmer error, raised as such; everything from validation
        # on is an evaluation failure, logged and classified.
        raw = self._raw_fields(input, fields)
        try:
            values = validate_inputs(raw, self.contract.input_schema)
            text = values.get(self._text_field, "") if self._text_field else ""
            grade_level = values.get("grade_level", "")
            self.logger.info(
                "Starting %s evaluation",
                self.metadata.label,
                extra={**context, "grade_level": grade_level, "text_length": len(text)},
            )

            messages = self._render_messages(values)
            dependency, model = provider_context(self.provider)
            assert self._step.model is not None
            response = await call_with_resampling(
                lambda: self.provider.generate_structured(
                    messages,
                    self.output_model,
                    temperature=self.effective_temperature(self._step.temperature),
                ),
                max_retries=self.config.max_retries,
                dependency=dependency,
                model=model,
                logger=self.logger,
            )

            elapsed_ms = int((time.perf_counter() - start) * 1000)
            result: EvaluationResult[Any] = EvaluationResult(
                evaluator=self.metadata.id,
                result=response.data,
                metadata=EvaluationMetadata(
                    model=self.provider.label,
                    processing_time_ms=elapsed_ms,
                    token_usage=EvaluationTokenUsage(
                        input_tokens=response.usage.input_tokens,
                        output_tokens=response.usage.output_tokens,
                    ),
                ),
            )
            outcome = self.metadata.outcome
            self.logger.info(
                "%s evaluation completed successfully",
                self.metadata.label,
                extra={
                    **context,
                    "grade_level": grade_level,
                    "score": getattr(response.data, outcome.score, None) if outcome else None,
                    "processing_time_ms": elapsed_ms,
                },
            )
            return cast(EvaluationResult[OutputT], result)
        except Exception as error:
            elapsed_ms = int((time.perf_counter() - start) * 1000)
            self.logger.error(
                "%s evaluation failed",
                self.metadata.label,
                extra={
                    **context,
                    "grade_level": grade_level,
                    "error": type(error).__name__,
                    "processing_time_ms": elapsed_ms,
                },
            )
            # Nothing is re-classified here. Every provider failure was already mapped by
            # ``call_with_resampling``; anything else reaching this point is a fault in this
            # SDK, and wrapping it as a provider error would name a service that did not
            # fail (spec §6.2) and hide the bug.
            raise

    # --- pieces of the flow ------------------------------------------------------------

    def _prompt_inputs(self, values: Mapping[str, str]) -> dict[str, str]:
        """Every placeholder the step declares, resolved from the source the contract names."""
        computed: dict[str, str] = {}
        for entry in self._preprocessing:
            if entry.condition is not None and not entry.condition.holds(dict(values)):
                continue
            assert entry.python is not None and entry.output is not None
            source = values.get(entry.input or "", "")
            computed[entry.output] = format_number(run_preprocessing_step(source, entry.python))

        assert self._step.prompt is not None
        inputs: dict[str, str] = {}
        for name, placeholder in self._step.prompt.placeholders.items():
            kind, _, rest = placeholder.source.partition(".")
            if placeholder.source == "input":
                value = values.get(name)
            elif kind == "input":
                value = values.get(rest)
            else:  # preprocessing.<output>
                value = computed.get(rest)
            if value is None:
                if placeholder.required:
                    # The contract's own inconsistency, not a service's: registry data is
                    # the caller's fault domain (spec §6.1), and stamping a provider on it
                    # would name a dependency that did not fail (§6.2).
                    raise ConfigurationError(
                        f'Placeholder "{name}" in {self.metadata.name} has no value '
                        f"from source {placeholder.source!r}."
                    )
                continue
            inputs[name] = value
        return inputs

    def _render_messages(self, values: Mapping[str, str]) -> list[Message]:
        assert self._step.prompt is not None
        inputs = self._prompt_inputs(values)
        placeholders = list(self._step.prompt.placeholders)
        return [
            Message(
                role=message.role,
                content=render_prompt(
                    self.contract.document(message.source_path), inputs, placeholders
                ),
            )
            for message in self._step.prompt.messages
        ]


__all__ = ["SingleStepEvaluator", "step_for"]
